packages/pages-shared/__tests__/asset-server/handler.test.ts

```diff
@@ -0,0 +1,258 @@
+import type { HandlerContext } from "../../asset-server/handler";
+import type { Metadata } from "../../asset-server/metadata";
+import type { RedirectRule } from "../../metadata-generator/types";
+
+describe("asset-server handler", () => {
+	test("Returns appropriate status codes", async () => {
+		const statuses = [301, 302, 303, 307, 308];
+		const metadata = createMetadataObjectWithRedirects(
+			statuses
+				.map((status) => ({
+					status,
+					from: `/${status}`,
+					to: "/home",
+				}))
+				.concat({
+					status: 302,
+					from: "/500",
+					to: "/home",
+				})
+		);
+
+		const tests = statuses.map(async (status) => {
+			const { response } = await getTestResponse({
+				request: "https://foo.com/" + status,
+				metadata,
+			});
+
+			expect(response.status).toBe(status);
+			expect(response.headers.get("Location")).toBe("/home");
+		});
+
+		await Promise.all(tests);
+
+		const { response } = await getTestResponse({
+			request: "https://foo.com/500",
+			metadata,
+		});
+
+		expect(response.status).toBe(302);
+		expect(response.headers.get("Location")).toBe("/home");
+	});
+
+	test("Match exact pathnames, before any HTML redirection", async () => {
+		const metadata = createMetadataObjectWithRedirects([
+			{ from: "/", to: "/home", status: 301 },
+			{ from: "/page.html", to: "/elsewhere", status: 301 },
+		]);
+		const findAssetEntryForPath = async (path: string) => {
+			if (path === "/index.html") {
+				return "index page";
+			}
+
+			if (path === "/page.html") {
+				return "some page";
+			}
+
+			return null;
+		};
+
+		{
+			const { response } = await getTestResponse({
+				request: "/index.html",
+				metadata,
+				findAssetEntryForPath,
+			});
+			expect(response.status).toBe(308);
+			expect(response.headers.get("Location")).toEqual("/");
+		}
+		{
+			const { response } = await getTestResponse({
+				request: "/index",
+				metadata,
+				findAssetEntryForPath,
+			});
+			expect(response.status).toBe(308);
+			expect(response.headers.get("Location")).toEqual("/");
+		}
+		{
+			const { response } = await getTestResponse({
+				request: "/",
+				metadata,
+				findAssetEntryForPath,
+			});
+			expect(response.status).toBe(301);
+			expect(response.headers.get("Location")).toEqual("/home");
+		}
+		{
+			// The redirect rule takes precedence to the 308s
+			const { response } = await getTestResponse({
+				request: "/page.html",
+				metadata,
+				findAssetEntryForPath,
+			});
+			expect(response.status).toBe(301);
+			expect(response.headers.get("Location")).toEqual("/elsewhere");
+		}
+		{
+			// This serves the HTML even though the redirect rule is present
+			const { response, spies } = await getTestResponse({
+				request: "/page",
+				metadata,
+				findAssetEntryForPath,
+			});
+			expect(response.status).toBe(200);
+			expect(spies.fetchAsset).toBe(1);
+			// expect(response.url).toBe("https://foo.com/page.html");
```

Comment:
This is coming out to be `http://example.com/asset` for some reason and I'm still figuring out why